pkg/utils/etcdutil/etcdutil.go

```diff
@@ -196,72 +210,237 @@ func EtcdKVPutWithTTL(ctx context.Context, c *clientv3.Client, key string, value
-	failpoint.Inject("autoSyncInterval", func() {
-		autoSyncInterval = 10 * time.Millisecond
-	})
-	failpoint.Inject("closeKeepAliveCheck", func() {
-		autoSyncInterval = 0
-		dialKeepAliveTime = 0
-		dialKeepAliveTimeout = 0
-	})
 	client, err := clientv3.New(clientv3.Config{
 		Endpoints:            endpoints,
 		DialTimeout:          defaultEtcdClientTimeout,
-		AutoSyncInterval:     autoSyncInterval,
 		TLS:                  tlsConfig,
 		LogConfig:            &lgc,
-		DialKeepAliveTime:    dialKeepAliveTime,
-		DialKeepAliveTimeout: dialKeepAliveTimeout,
+		DialKeepAliveTime:    defaultDialKeepAliveTime,
+		DialKeepAliveTimeout: defaultDialKeepAliveTimeout,
 	})
-	if err == nil {
-		log.Info("create etcd v3 client", zap.Strings("endpoints", endpoints))
+	return client, err
+}
+
+// CreateEtcdClient creates etcd v3 client with detecting endpoints.
+func CreateEtcdClient(tlsConfig *tls.Config, acURLs []url.URL) (*clientv3.Client, error) {
+	urls := make([]string, 0, len(acURLs))
+	for _, u := range acURLs {
+		urls = append(urls, u.String())
 	}
+	client, err := newClient(tlsConfig, urls...)
+	if err != nil {
+		return nil, err
+	}
+
+	tickerInterval := defaultDialKeepAliveTime
+	failpoint.Inject("fastTick", func() {
+		tickerInterval = 100 * time.Millisecond
+	})
+	failpoint.Inject("closeTick", func() {
+		tickerInterval = 0
+	})
+	if tickerInterval == 0 {
+		return client, err
+	}
+
+	checker := &healthyChecker{
+		tlsConfig: tlsConfig,
+	}
+	eps := syncUrls(client)
+	checker.update(eps)
+
+	// Create a goroutine to check the health of etcd endpoints periodically.
+	go func(client *clientv3.Client) {
+		defer logutil.LogPanic()
+		ticker := time.NewTicker(tickerInterval)
+		defer ticker.Stop()
+		lastAvailable := time.Now()
+		for {
+			select {
+			case <-client.Ctx().Done():
+				log.Info("[etcd client] etcd client is closed, exit health check goroutine")
+				checker.Range(func(key, value interface{}) bool {
+					client := value.(*healthyClient)
+					client.Close()
+					return true
+				})
+				return
+			case <-ticker.C:
+				usedEps := client.Endpoints()
+				healthyEps := checker.patrol(client.Ctx())
+				if len(healthyEps) == 0 {
+					// when all endpoints are unhealthy, try to reset endpoints to update connect
+					// rather than delete them to avoid there is no any endpoint in client.
+					if time.Since(lastAvailable) > etcdServerDisconnectedTimeout {
+						log.Info("[etcd client] no available endpoint, try to reset endpoints", zap.Strings("last-endpoints", usedEps))
+						client.SetEndpoints([]string{}...)
+						client.SetEndpoints(usedEps...)
+					}
+				} else {
+					if !isEqual(healthyEps, usedEps) {
+						client.SetEndpoints(healthyEps...)
+						change := fmt.Sprintf("%d->%d", len(usedEps), len(healthyEps))
+						etcdStateGauge.WithLabelValues("endpoints").Set(float64(len(healthyEps)))
+						log.Info("[etcd client] update endpoints", zap.String("num-change", change),
```

Comment:
I think just printing them is ok?

Reply:
The number of etcd servers is three to five that I don't think printing them all would cause any additional trouble

Reply:
I mean not just print the number but print the addresses?

Reply:
how about printing diff of addresses rather than all the last addresses and current addresses?

Reply:
I think it's ok to print them all if the operation is not frequent.

Reply:
ok, it only prints when endpoints changed, for example, network partition and recover, pd leader io hang and recover, etcd scale-in or scale-out.

Reply:
Is it necessary now?

Reply:
yes, it is.